### PR TITLE
[Bug Fix] Fix state sync stuck in infinite loop discovering the next bad peer

### PR DESCRIPTION
**Pull request by yzang2019**
## Describe your changes and provide context
**Problem Statement:**
1. State sync requires to have at least 2 valid snapshot peers to start with
2. The default peer dialing order is Unconditional peers (score 255) -> Persistent peers (254)-> other peers order by score (0)
3. If there is a bad peer which pass dial, however always fail to receive message, tendermint will disconnect with that peer, and it will still be the next peer to dial in the loop (since we would never update the score or change the ordering of dial)

**Solution:**
1. We fixed the scoring system for peers by introducing a default score (10) instead of 0, and we will decrease the score for peers that frequently disconnected or dial failed
2. We fixed the cache logic so that each time score is updated, we will invalidate the cache and the ordering will always reflecting with latest score 

## Testing performed to validate your change
Tested on RPC node and verified it works:
root@ip-172-31-2-185:/home/ubuntu# curl localhost:26657/net_info |jq
> "peer_connections": [
>     {
>       "node_id": "ccef4d9dab0c12ea820ecefafb728a5d8eb39297",
>       "state": "ready,connected",
>       "score": "254"
>     },
>     {
>       "node_id": "79389ef8775ad3310b77fcd935db30f32b5ba764",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "1d29a2989fbee9b10cf93942a27408563a48929b",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "c77ecbb674b29e47dc2e979bc42fd48c8ab0440b",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "650a118a5919c1d0eb3d9f17b14cfb2a6b1c8b9d",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "06a80e8564b06ea5afef5ae74face1798e505572",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "4944c0fb34a76ad537f4eefa1734d6f6a2da5ed0",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "1d3e3c08e5a39e7f211a1b256739aa2a3704e9e1",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "20b4f9207cdc9d0310399f848f057621f7251846",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "802bd5bebded432a15e83a5cd74a4fb4a8e5a576",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "862b03573172a3366afe1cabb903ba0552689e63",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "6090d4c2e2f6e70718266900e3cb4b83ffa209ea",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "dd7b2d3e981aed758f11240e2063462ef1cb913b",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "d67c2434027ec24c0d907b8ff663174d8e9916bc",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "45486ab4af9dd54560b402d2f8f9bbedcf230ba1",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "42811a231aa0337cde84e5961b411e0bdb5179cf",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "65c257f9275beb1b99ca169ef89743c034b15db0",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "8850005bbfac49fa98850bd2eb27f7631a33031f",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "6bfeb8d02f6a01c492c8d169b1d53ede238bff17",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "56a1d17ff164627a1102528014d4d165f9862985",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "ba9f16767d2f695c97111c17ca2dd915eedc77da",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "619d116d9a20c1ee58d9844f4ad44fd635f1c847",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "0dcce954ad68259aa376d2a0cdc6581d74c0fa12",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "170397e75ca2b0f4e9f3b1bb5d0d23f9b10f01c7",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "5710d992d9c33b01c3b23df4cbd715e9b4c7b46b",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "f95c7153905d673a923c53defdcbe57145b95a53",
>       "state": "ready,connected",
>       "score": "10"
>     },
>     {
>       "node_id": "8f61c476ae8862cf5a965f4cb61eb5e217b61927",
>       "state": "ready,connected",
>       "score": "9"
>     },
>     {
>       "node_id": "965c421affb504e5a6095817303e1eb450f20b95",
>       "state": "",
>       "score": "9"
>     },
>     {
>       "node_id": "37bee40881a13a056d978b3228d9efbae8dbb6fa",
>       "state": "ready,connected",
>       "score": "9"
>     },
>     {
>       "node_id": "622edfc381a73cb9a624815831d3cbfecab04e4a",
>       "state": "ready,connected",
>       "score": "9"
>     },
>     {
>       "node_id": "a62894db2e6b6ca0f39a5c82693f8837122ab4a1",
>       "state": "",
>       "score": "8"
>     },
>     {
>       "node_id": "dfedc0a045e85c6d6938a8483681cc12dcd87fe3",
>       "state": "ready,connected",
>       "score": "8"
>     },
>     {
>       "node_id": "2c9ee2596021eabc24e9501389f280dab62b3b30",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "c6daa0350abb0ac0ed2ce65d15da04800f7e139e",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "83710b3e5ec49ffe61418f4c89845befba8d1091",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "955f7c2e13df7af4ffde385e9e7c68634300e6aa",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "9d4859b0081303961e814016749a2d26f51d7e8f",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "78bcfe7a2b24b3c68b7c2e4e6388cd41684213de",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "049629c1983b73f0a8076c9045c87289aa4c4f92",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "ac19a5c7236ac36bc9c74644901706fa71d81459",
>       "state": "",
>       "score": "7"
>     },
>     {
>       "node_id": "6947733fbac29f65e624e007c428e490dcaf4d8a",
>       "state": "",
>       "score": "7"
>     }
>   ]
>

### CODE REVIEW

File: internal/p2p/peermanager.go [hunk 1]
```diff
@@ -433,6 +435,7 @@ func (m *PeerManager) Add(address NodeAddress) (bool, error) {
 
 	// else add the new address
 	peer.AddressInfo[address] = &peerAddressInfo{Address: address}
+	fmt.Printf("Adding new peer %s with address %s to peer store\n", peer.ID, address.String())
```
Comment (philipsu522) [inline internal/p2p/peermanager.go:438]: nit can we use logger instead of fmt?

File: internal/p2p/router.go [hunk 2]
```diff
@@ -641,9 +642,9 @@ func (r *Router) connectPeer(ctx context.Context, address NodeAddress) {
 	case errors.Is(err, context.Canceled):
 		return
 	case err != nil:
-		r.logger.Error("failed to dial peer", "peer", address, "err", err)
+		r.logger.Debug("failed to dial peer", "peer", address, "err", err)
```
Comment (philipsu522) [inline internal/p2p/router.go:645]: nice

**Review by philipsu522 — APPROVED** [review]

File: internal/p2p/peermanager.go (hunk 1 above)
Comment (yzang2019) [reply internal/p2p/peermanager.go:438]: It seems peermanager doesn't support logger yet, let me check if it's easy to add

Comment (yzang2019) [reply internal/p2p/peermanager.go]: Added the change to enable logger for peer manager

**Review by philipsu522 — APPROVED** [review]

**Review by philipsu522 — APPROVED** [review]

**Pull request merged**